### ISSUE TITLE
Different slider stop value on Python 3.11

### ISSUE THREAD

**Issue by LucaMarconato** [napari/napari#7208]
### 🐛 Bug Report

The slider end was `n_channel - 1` for Python < 3.11, but is `n_channel` for Python 3.11.

Note, discovered thanks to this CI run: https://github.com/scverse/napari-spatialdata/pull/301.

### 💡 Steps to Reproduce

```python
import numpy as np
import napari

# Create random data for a multichannel image with 5 channels
data = np.random.random((512, 512, 5))

viewer = napari.Viewer()
viewer.add_image(data)

# Access the start, stop, step values of the slider for the channel dimension
print(viewer.dims.range)
start, stop, step = viewer.dims.range[2]
print(f"Slider range: start={start}, stop={stop}, step={step}")

# Check if the slider is present and assert the values
assert start == 0
assert step == 1
print(stop)
# the assertion fails for Python 3.11
assert stop == 4
```

### 💡 Expected Behavior

The assertion should not fail, as in previous versions of Python.

### 🌎 Environment
```
napari: 0.4.19.post1
Platform: macOS-14.6.1-arm64-arm-64bit
System: MacOS 14.6.1
Python: 3.11.0 | packaged by conda-forge | (main, Jan 14 2023, 12:26:40) [Clang 14.0.6 ]
Qt: 5.15.14
PyQt5: 5.15.11
NumPy: 1.26.4
SciPy: 1.14.0
Dask: 2024.8.1
VisPy: 0.14.3
magicgui: 0.9.1
superqt: 0.6.7
in-n-out: 0.2.1
app-model: 0.2.8
npe2: 0.7.7

OpenGL:
  - GL version:  2.1 Metal - 88.1
  - MAX_TEXTURE_SIZE: 16384

Screens:
  - screen 1: resolution 3840x2160, scale 1.0

Settings path:
  - /Users/macbook/Library/Application Support/napari/ome311_5c99b3d35ca8ee4c9df9ed01b2d1118f174296d3/settings.yaml
Plugins:
  - napari: 0.4.19.post1 (77 contributions)
  - napari-console: 0.0.9 (0 contributions)
  - napari-matplotlib: 2.0.3 (10 contributions)
  - napari-spatialdata: 0.5.2.post2.dev6+ge9dd1d7 (8 contributions)
  - napari-svg: 0.2.0 (2 contributions)
```

### 💡 Additional Context

The problem that I report is not a blocker for our case.

**Comment by psobolewskiPhD** [comment]
Very odd. I can indeed reproduce with python 3.11 and napari 0.4.19post1 like the OP:
```
napari: 0.4.19.post1
Platform: macOS-14.5-arm64-arm-64bit
System: MacOS 14.5
Python: 3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:34:54) [Clang 16.0.6 ]
Qt: 5.15.14
PyQt5: 5.15.11
NumPy: 1.26.4
SciPy: 1.14.0
Dask: 2024.8.1
VisPy: 0.14.3
magicgui: 0.9.1
superqt: 0.6.7
in-n-out: 0.2.1
app-model: 0.2.8
npe2: 0.7.7

OpenGL:
  - GL version:  2.1 Metal - 88.1
  - MAX_TEXTURE_SIZE: 16384

Screens:
  - screen 1: resolution 2056x1329, scale 2.0
  - screen 2: resolution 2560x1440, scale 1.0

Settings path:
  - /Users/sobolp/Library/Application Support/napari/napari-052-11_4a88921f87d40c342d8e10938499709281cacd2b/settings.yaml
Plugins:
  - napari: 0.4.19.post1 (77 contributions)
  - napari-console: 0.0.9 (0 contributions)
  - napari-svg: 0.2.0 (2 contributions)
```

If I update napari (pip install napari -U) in place to 0.5.2 then the assert passes.

**Comment by psobolewskiPhD** [comment]
Interestingly, the assert also fails with python 3.10 and napari 0.4.19post1:
```
napari: 0.4.19.post1
Platform: macOS-14.5-arm64-arm-64bit
System: MacOS 14.5
Python: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:51:49) [Clang 16.0.6 ]
Qt: 5.15.14
PyQt5: 5.15.11
NumPy: 1.26.4
SciPy: 1.14.0
Dask: 2024.8.1
VisPy: 0.14.3
magicgui: 0.9.1
superqt: 0.6.7
in-n-out: 0.2.1
app-model: 0.2.8
npe2: 0.7.7

OpenGL:
  - GL version:  2.1 Metal - 88.1
  - MAX_TEXTURE_SIZE: 16384

Screens:
  - screen 1: resolution 2056x1329, scale 2.0
  - screen 2: resolution 2560x1440, scale 1.0

Settings path:
  - /Users/sobolp/Library/Application Support/napari/napari-052-10_ecae565493ee57ffa2f1c1fe8d98cb0ccdc6d7e7/settings.yaml
Plugins:
  - napari: 0.4.19.post1 (77 contributions)
  - napari-console: 0.0.9 (0 contributions)
  - napari-svg: 0.2.0 (2 contributions)
```
@LucaMarconato Can you post a `napari --info` where this passes for you? It seems like

**Comment by psobolewskiPhD** [comment]
Bisecting this is a nightmare 👿

**Comment by LucaMarconato** [comment]
> @LucaMarconato Can you post a napari --info where this passes for you?
> It doesn't seem to be python version related to me, but napari version related.

Here is the output of `napari --info`. It seems indeed to be related to the napari version.

```
napari: 0.5.2
Platform: macOS-14.6.1-arm64-arm-64bit
System: MacOS 14.6.1
Python: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:01:00)  [Clang 13.0.1 ]
Qt: 5.15.14
PyQt5: 5.15.11
NumPy: 1.26.4
SciPy: 1.13.1
Dask: 2024.8.0
VisPy: 0.14.3
magicgui: 0.9.1
superqt: 0.6.7
in-n-out: 0.2.1
app-model: 0.2.8
npe2: 0.7.7

OpenGL:
  - GL version:  2.1 Metal - 88.1
  - MAX_TEXTURE_SIZE: 16384
  - GL_MAX_3D_TEXTURE_SIZE: 2048

Screens:
  - screen 1: resolution 2056x1329, scale 2.0

Optional:
  - numba: 0.60.0
  - triangle not installed
  - napari-plugin-manager not installed

Settings path:
  - /Users/macbook/Library/Application Support/napari/ome39_6000391d6db035eb691da0b5944978e9012f7d97/settings.yaml
Plugins:
  - napari: 0.5.2 (81 contributions)
  - napari-console: 0.0.9 (0 contributions)
  - napari-matplotlib: 2.0.1 (10 contributions)
  - napari-spatialdata: 0.5.2.post2.dev4+ge1c05ae (8 contributions)
  - napari-svg: 0.2.0 (2 contributions)
```

**Comment by Czaki** [comment]
Yes. It was changed between napari 0.4.19 and 0.5.0 by @brisvag 
I do not remember motivation, but Know that need to handle this at some moment.

**Comment by brisvag** [comment]
I'm confused... `dims.range` used to be "somewhat analogous" to python's builtin `range`, with non-inclusive `stop`. This was a weird mix of integer and float logic, which we dropped when we moved to using world space coordinates as our source of truth in `Dims` (started in https://github.com/napari/napari/pull/5522 and continued in https://github.com/napari/napari/pull/5751). This was a breaking change in 0.5.0.

If I understand correctly, the `assert` above assumes the latter, which means it was designed with napari `0.5.0` and not `0.4.19`.

**Comment by psobolewskiPhD** [comment]
In my testing I couldn't find any python version effect, just the difference between 0.4.19 and 0.5.0.
So if we all agree that the 0.5.0 behavior is the correct one, I think we can close this?

**Comment by brisvag** [comment]
Depends, if @LucaMarconato expected/had this behaviour in 0.4.19 then we have some issue :p

**Comment by LucaMarconato** [comment]
I am fine with both behavior (so the one implemented in 0.5.0 looks good 😊).

**Comment by brisvag** [comment]
Ok, let's close this then!
